kdump: Add SUSE kdump config support

skazi0 opened the pull request:
If parsing /etc/kdump.conf doesn't return usable settings, try
/etc/sysconfig/kdump which is used by SUSE distributions as the main
kdump config.

The file is in dotenv format and currently only KDUMP_SAVEDIR,
KDUMP_DUMPFORMAT and KDUMP_SSH_IDENTITY entries are used by Cockpit.

SUSE supports additional dump target types (ftp, sftp, cifs) but doesn't
support others (raw, mount). The target dialog currently supports the
common types (nfs, ssh, local).

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+     * }
+     * skipNotify: Don't notify about changes, e.g.to avoid multiple updates when writing a file
+     */
+    _parseText(rawContent, skipNotify = false) {
+        this._dataAvailableResolve();
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+    _parseText(rawContent, skipNotify = false) {
+        this._dataAvailableResolve();
+
+        // clear settings if file is empty/missing
+        if (!rawContent) {
+            this._originalSettings = null;
+            this.settings = null;
+            if (!skipNotify)
+                this.dispatchEvent("kdumpConfigChanged", this.settings);
+            return;
+        }
```
These 8 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            return;
+        }
+
+        // if nothing changed, don't bother parsing the content
+        if (rawContent == this._rawContent)
+            return;
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+        if (rawContent == this._rawContent)
+            return;
+
+        this._rawContent = rawContent;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+        this._rawContent = rawContent;
+
+        // this is the format expected by the UI
+        this.settings = {
+            _internal: {},
+            targets: {},
+            compression: { enabled: false, allowed: true },
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            compression: { enabled: false, allowed: true },
+        };
+
+        this._lines = rawContent.split(/\r?\n/);
+        this._lines.forEach((line, index) => {
+            const trimmed = line.trim();
+            // if the line is empty or only a comment, skip
+            if (trimmed.indexOf("#") === 0 || trimmed.length === 0)
+                return;
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            if (trimmed.indexOf("#") === 0 || trimmed.length === 0)
+                return;
+
+            // parse KEY=value or KEY="value" line
+            let parts = trimmed.match(/^([A-Z_]+)=(.*)$/);
+            if (parts === null)
+                return;
+            const key = parts[1];
+            let value = parts[2];
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            const key = parts[1];
+            let value = parts[2];
+
+            // value might be quoted
+            let quoted = false;
+            if (value.startsWith("\"")) {
+                quoted = true;
+                parts = value.match(/^"([^"]*)"\s*(.*)$/);
+                // malformed line, no ending quote?
+                if (parts === null) {
+                    console.error("Malformed kdump config line:", line, "in", this.filename);
+                    return;
+                }
+            } else {
+                // not quoted should be simple value
+                parts = value.match(/^(\S+)\s*(.*)$/);
+                if (parts === null)
+                    parts = "";
+            }
+            value = parts[1];
+            const suffix = (parts[2] || "").trim();
```
These 18 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            value = parts[1];
+            const suffix = (parts[2] || "").trim();
+
+            this.settings._internal[key] = {
+                index: index,
+                value: value,
+                origLine: line,
+                quoted: quoted,
+                suffix: suffix
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                quoted: quoted,
+                suffix: suffix
+            };
+        });
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            };
+        });
+
+        // make sure we copy the original keys so we overwrite the correct lines when saving
+        this._originalSettings = { };
+        Object.keys(this.settings._internal).forEach((key) => {
+            this._originalSettings[key] = { ...this.settings._internal[key] };
+        });
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            this._originalSettings[key] = { ...this.settings._internal[key] };
+        });
+
+        this._extractSettings();
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+        this._extractSettings();
+
+        if (!skipNotify)
+            this.dispatchEvent("kdumpConfigChanged", this.settings);
+    }
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+    /* extract settings managed by cockpit from _internal into platform independent model
+     */
+    _extractSettings() {
+        // generate target(s) from KDUMP_SAVEDIR
+        if ("KDUMP_SAVEDIR" in this.settings._internal && this.settings._internal.KDUMP_SAVEDIR.value) {
+            let savedir = this.settings._internal.KDUMP_SAVEDIR.value;
+            // handle legacy "file" without prefix
+            if (savedir.startsWith("/"))
+                savedir = "file://" + savedir;
+            // server includes "username:password@" and can be empty for file://
+            const parts = savedir.match(/^(.*):\/\/([^/]*)(\/.*)$/);
+            // malformed KDUMP_SAVEDIR
+            if (parts === null) {
+                console.error("Malformed KDUMP_SAVEDIR entry:", savedir, "in", this.filename);
+                return;
+            }
+            const [, scheme, server, path] = parts;
+            if (scheme === "file") {
+                this.settings.targets.local = {
+                    type: "local",
+                    path: path,
```
These 19 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                    type: "local",
+                    path: path,
+                };
+            } else if (scheme === "nfs") {
+                this.settings.targets.nfs = {
+                    type: scheme,
+                    // on read full path is used as export
+                    export: path,
+                    server: server,
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                    export: path,
+                    server: server,
+                };
+            } else {
+                this.settings.targets[scheme] = {
+                    type: scheme,
+                    path: path,
+                    server: server,
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                    path: path,
+                    server: server,
+                };
+                // sshkey is used by ssh and sftp/scp
+                if ("KDUMP_SSH_IDENTITY" in this.settings._internal) {
+                    this.settings.targets[scheme].sshkey =
+                        this.settings._internal.KDUMP_SSH_IDENTITY.value;
+                }
+            }
+        }
```
These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            }
+        }
+
+        // default to local if no target configured
+        if (Object.keys(this.settings.targets).length === 0)
+            this.settings.targets.local = { type: "local" };
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+        if (Object.keys(this.settings.targets).length === 0)
+            this.settings.targets.local = { type: "local" };
+
+        this.settings.compression.enabled = (
+            !("KDUMP_DUMPFORMAT" in this.settings._internal) ||
+            // TODO: what about other compression formats (lzo, snappy)?
+            this.settings._internal.KDUMP_DUMPFORMAT.value === "compressed"
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            // TODO: what about other compression formats (lzo, snappy)?
+            this.settings._internal.KDUMP_DUMPFORMAT.value === "compressed"
+        );
+    }
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+     * make sure setting exists if value is not empty
+     * don't delete existing settings
+     */
+    _updateSetting(settings, key, value) {
+        if (key in settings._internal) {
+            settings._internal[key].value = value;
+        } else {
+            if (value)
+                settings._internal[key] = { value: value };
+        }
+    }
```
These 8 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+     * this.settings = current state from file
+     * settings = in-memory state from UI
+     */
+    _persistSettings(settings) {
+        // target
+        if (Object.keys(settings.targets).length > 0) {
+            const target = Object.values(settings.targets)[0];
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+        if (Object.keys(settings.targets).length > 0) {
+            const target = Object.values(settings.targets)[0];
+
+            this._updateSetting(settings, "KDUMP_SSH_IDENTITY", target.sshkey);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+            this._updateSetting(settings, "KDUMP_SSH_IDENTITY", target.sshkey);
+
+            let savedir;
+            // default for empty path (except nfs, see below)
+            let path = target.path || "/var/crash";
+            if (path && !path.startsWith("/"))
+                path = "/" + path;
+            if (target.type === "local") {
+                savedir = "file://" + path;
+            } else if (target.type === "nfs") {
+                // override empty path default as nfs path is merged into export on read
+                if (!target.path)
+                    path = "";
+                let exprt = target.export;
+                if (!exprt.startsWith("/"))
+                    exprt = "/" + exprt;
+                savedir = "nfs://" + target.server + exprt + path;
+            } else {
+                savedir = target.type + "://" + target.server + path;
+            }
+            this._updateSetting(settings, "KDUMP_SAVEDIR", savedir);
+        }
+        // compression
+        if (this.settings.compression.enabled != settings.compression.enabled) {
+            if (settings.compression.enabled) {
+                this._updateSetting(settings, "KDUMP_DUMPFORMAT", "compressed");
+            } else {
+                this._updateSetting(settings, "KDUMP_DUMPFORMAT", "ELF");
+            }
+        }
+        return settings;
+    }
```
These 30 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+    /* generate the config file from raw text and settings
+     */
+    _generateConfig(settings) {
+        settings = this._persistSettings(settings);
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+    _generateConfig(settings) {
+        settings = this._persistSettings(settings);
+
+        const lines = this._lines.slice(0);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+        const lines = this._lines.slice(0);
+
+        // we take the lines from our last read operation and modify them with the new settings
+        Object.keys(settings._internal).forEach((key) => {
+            const entry = settings._internal[key];
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+        Object.keys(settings._internal).forEach((key) => {
+            const entry = settings._internal[key];
+
+            let value = entry.value !== undefined ? entry.value : "";
+            // quote what was quoted before + empty values + multi-word values
+            if (entry.quoted || value === "" || value.includes(" "))
+                value = "\"" + value + "\"";
+            let line = key + "=" + value;
+            if (entry.suffix)
+                line = line + " " + entry.suffix;
+            // this might be a new entry
+            if (!(key in this._originalSettings)) {
+                lines.push(line);
+                return;
+            }
+            // otherwise edit the old line
+            const origEntry = this._originalSettings[key];
+            lines[origEntry.index] = line;
+        });
```
These 16 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            lines[origEntry.index] = line;
+        });
+
+        // make sure file ends with EOL
+        if (lines[lines.length - 1] !== "")
+            lines.push("");
+        return lines.join("\n");
+    }
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/kdump-client.js:
```diff
@@ -61,6 +62,19 @@ export class KdumpClient {
+
+        this.configClient.wait().then(() => {
+            // if no configuration found, try SUSE version
+            if (this.configClient.settings === null) {
+                this.configClient.close();
+                this.configClient = new ConfigFileSUSE("/etc/sysconfig/kdump", true);
+                this._watchConfigChanges();
+            }
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/kdump-view.jsx:
```diff
@@ -364,6 +364,12 @@ export class KdumpPage extends React.Component {
                     /* mount targets outside of nfs are too complex for the
                      * current target dialog */
                     kdumpLocation = _("On a mounted device");
+                } else if (target.type == "ftp") {
+                    kdumpLocation = _("Remote over FTP");
+                } else if (target.type == "sftp") {
+                    kdumpLocation = _("Remote over SFTP");
+                } else if (target.type == "cifs") {
+                    kdumpLocation = _("Remote over CIFS/SMB");
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220810-111027-b2d4a356-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

skazi0 commented:
rebased to get https://github.com/cockpit-project/cockpit/pull/17652

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            // parse KEY=value or KEY="value" line
+            let parts = trimmed.match(/^([A-Z_]+)=(.*)$/);
+            if (parts === null)
+                return;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                quoted = true;
+                parts = value.match(/^"([^"]*)"\s*(.*)$/);
+                // malformed line, no ending quote?
+                if (parts === null) {
+                    console.error("Malformed kdump config line:", line, "in", this.filename);
+                    return;
+                }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                // not quoted should be simple value
+                parts = value.match(/^(\S+)\s*(.*)$/);
+                if (parts === null)
+                    parts = "";
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            let savedir = this.settings._internal.KDUMP_SAVEDIR.value;
+            // handle legacy "file" without prefix
+            if (savedir.startsWith("/"))
+                savedir = "file://" + savedir;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            // server includes "username:password@" and can be empty for file://
+            const parts = savedir.match(/^(.*):\/\/([^/]*)(\/.*)$/);
+            // malformed KDUMP_SAVEDIR
+            if (parts === null) {
+                console.error("Malformed KDUMP_SAVEDIR entry:", savedir, "in", this.filename);
+                return;
+            }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+     * don't delete existing settings
+     */
+    _updateSetting(settings, key, value) {
+        if (key in settings._internal) {
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+    _updateSetting(settings, key, value) {
+        if (key in settings._internal) {
+            settings._internal[key].value = value;
+        } else {
+            if (value)
+                settings._internal[key] = { value: value };
+        }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                    path = "";
+                let exprt = target.export;
+                if (!exprt.startsWith("/"))
+                    exprt = "/" + exprt;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+        }
+        // compression
+        if (this.settings.compression.enabled != settings.compression.enabled) {
+            if (settings.compression.enabled) {
+                this._updateSetting(settings, "KDUMP_DUMPFORMAT", "compressed");
+            } else {
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                value = "\"" + value + "\"";
+            let line = key + "=" + value;
+            if (entry.suffix)
+                line = line + " " + entry.suffix;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            if (entry.suffix)
+                line = line + " " + entry.suffix;
+            // this might be a new entry
+            if (!(key in this._originalSettings)) {
+                lines.push(line);
+                return;
+            }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/kdump-view.jsx:
```diff
@@ -364,6 +364,12 @@ export class KdumpPage extends React.Component {
                     /* mount targets outside of nfs are too complex for the
                      * current target dialog */
                     kdumpLocation = _("On a mounted device");
+                } else if (target.type == "ftp") {
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220824-154033-c3fadb1f-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                // not quoted should be simple value
+                parts = value.match(/^(\S+)\s*(.*)$/);
+                if (parts === null)
+                    parts = ["", ""];
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            let savedir = this.settings._internal.KDUMP_SAVEDIR.value;
+            // handle legacy "file" without prefix
+            if (savedir.startsWith("/"))
+                savedir = "file://" + savedir;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            // server includes "username:password@" and can be empty for file://
+            const parts = savedir.match(/^(.*):\/\/([^/]*)(\/.*)$/);
+            // malformed KDUMP_SAVEDIR
+            if (parts === null) {
+                console.error("Malformed KDUMP_SAVEDIR entry:", savedir, "in", this.filename);
+                return;
+            }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                    path = "";
+                let exprt = target.export;
+                if (!exprt.startsWith("/"))
+                    exprt = "/" + exprt;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220829-200531-97830531-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+            // value might be quoted
+            let quoted = false;
+            if (value.startsWith("\"")) {
```
This is easier to read as `'"'` (i.e. using single quotes)

skazi0 replied:
done

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                return;
+
+            // parse KEY=value or KEY="value" line
+            let parts = trimmed.match(/^([A-Z_]+)=(.*)$/);
```
This doesn't allow spaces around the `=`, is that intentional? As this is a human editable config file, it should be robust against that.

More precisely, it doesn't allow it before the `=`, and spaces after the `=` would be part of the value, which is often not intended.

skazi0 replied:
The manpage says:
```
       The file format consists of shell variables (which means that the configuration file is actually parsed by a sh(1) compatible shell) that are described below. It’s recommended to use a pair of double quotes to assign values to that variables.
```
so there should be no spaces around the `=` sign.
Some of our tools support that but write modifications without spaces.
I'll add the same behavior.

skazi0 replied:
done

martinpitt replied:
OK, so it accepts spaces now, even though they should be illegal. But that seems fine, as long as it writes the file correctly.

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+
+            // value might be quoted
+            let quoted = false;
+            if (value.startsWith("\"")) {
```
If the line is `key = "value"`, this doesn't work with the current PR, as the first char of value is the space.

skazi0 replied:
should work now

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                }
+            } else {
+                // not quoted should be simple value
+                parts = value.match(/^(\S+)\s*(.*)$/);
```
A line line `key = foo bar` would come out as `settings["key"] == "foo"`, is that really intended? This seems very unintuitive. Wouldn't it make more sense to strip away `#.*$` first, strip the rest, and take that as value?

Or is it important that this file behaves like shell commands? But that would also not simply ignore the second word, that would be a syntax error.

skazi0 replied:
Yeah, this should be correct shell variable, but again our tools seem to accept that and fix on write. I will follow this.

skazi0 replied:
done

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+                    parts = ["", ""];
+            }
+            value = parts[1];
+            const suffix = (parts[2] || "").trim();
```
can parts[2] ever be something else than a string? I don't see how.

skazi0 replied:
This is a leftover. The `parts = ["", ""]` was added later. I'll clean this up.

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            let value = entry.value !== undefined ? entry.value : "";
+            // quote what was quoted before + empty values + multi-word values
+            if (entry.quoted || value === "" || value.includes(" "))
+                value = "\"" + value + "\"";
```
Again, use single quotes for strings which contain ".

skazi0 replied:
done

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+            lines[origEntry.index] = line;
+        });
+
+        // make sure file ends with EOL
```
s/EOL/newline/

skazi0 replied:
done

martinpitt commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,262 @@
+/* Parse an dotenv-style config file
+ * and monitor it for changes
+ */
+export class ConfigFileSUSE extends ConfigFile {
```
Perhaps it makes more sense to call this ConfigFileSysconfig (and the file accordingly)? Or is this SUSE specific because of the supported backends?

skazi0 replied:
This is pretty specific for SUSE. Note that RedHat also has `/etc/sysconfig/kdump` but it's not used for target config etc.

martinpitt commented on pkg/kdump/kdump-client.js:
```diff
@@ -61,6 +62,19 @@ export class KdumpClient {
+        this._watchConfigChanges();
+
+        this.configClient.wait().then(() => {
+            // if no configuration found, try SUSE version
+            if (this.configClient.settings === null) {
```
This is why I suggested to name the file more generically. There is no OS detection for "SUSE" here whatsoever, it's trying the sysconfig format everywhere.

skazi0 replied:
I'm fine with more generic name this but it might be confusing for other distros which use `/etc/sysconfig/kdump` but have different format inside.

martinpitt replied:
I see; let's leave it as is then, thanks!

martinpitt commented on test/verify/check-kdump:
```diff
@@ -248,6 +248,109 @@ class TestKdump(KdumpHelpers):
+
+        # Check remote CIFS location (no config dialog)
+        m.execute("sed -i 's/KDUMP_SAVEDIR=.*/KDUMP_SAVEDIR=cifs:\\/\\/user:pass@smbserver\\/dumps3/g' /etc/sysconfig/kdump")
+        b.wait_text("#kdump-target-info", "Remote over CIFS/SMB")
```
Can you please also test some "malformed config file" cases here? Such as a line not having a `=`, spaces around the `=`, or an unquoted multi-word value? These should just cause a console error and do something sensible in the UI.

martinpitt replied:
I didn't even mean to cover this console error, just the UI behaviour. But in fact you can: `self.browser.get_js_log()` gives you a list of log messages.

testlib.py uses that to check for messages with "error:", and fails the test if there are any. So that test would likely fail once you add such a case. You can explicitly allow it with calling `self.allow_browser_errors()`. *However*: errors should really only be logged for internal programming errors; for such external errors you should rather use console.warn().

skazi0 replied:
I changed errors to warnings and added some tests around the malformed inputs.

martinpitt reviewed (changes requested):
Thanks, great work! I have some concerns and questions around the parsing, but these shouldn't be too hard to sort out.

skazi0 commented:
@martinpitt I added some tests and changed parsing to cover special cases you mentioned. I put it into a separate commit for easier review.
/me needs to remember to clean this up before merging

martinpitt commented:
[this is an unrelated failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-051857-1fd368e9-debian-testing/log.html#155-2), similar to [this one](https://cockpit-logs.us-east-1.linodeobjects.com/pull-3807-20220901-081216-73c8d5e2-fedora-37-cockpit-project-cockpit/log.html#74-2). I've seen it a few times recently. @mvollmer, something for you? :man_pilot: Retrying.

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+        this._dataAvailableResolve();
+
+        // clear settings if file is empty/missing
+        if (!rawContent) {
+            this._originalSettings = null;
+            this.settings = null;
+            if (!skipNotify)
+                this.dispatchEvent("kdumpConfigChanged", this.settings);
+            return;
+        }
```
These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 7 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+
+        // if nothing changed, don't bother parsing the content
+        if (rawContent == this._rawContent)
+            return;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                quoted = true;
+                parts = value.match(/^"([^"]*)"\s*(.*)$/);
+                // malformed line, no ending quote?
+                if (parts === null) {
+                    console.warn("Incorrectly quoted value in kdump config line:", line, "in", this.filename);
+                    return;
+                }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                // not quoted should be simple value but grab everything and quote on write
+                parts = value.match(/^([^#]+?)\s*(#.*)?$/);
+                if (parts === null)
+                    parts = ["", ""];
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous replied:
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+
+    /* extract settings managed by cockpit from _internal into platform independent model
+     */
+    _extractSettings() {
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+            let savedir = this.settings._internal.KDUMP_SAVEDIR.value;
+            // handle legacy "file" without prefix
+            if (savedir.startsWith("/"))
+                savedir = "file://" + savedir;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                console.warn("Malformed KDUMP_SAVEDIR entry:", savedir, "in", this.filename);
+                return;
+            }
+            const [, scheme, server, path] = parts;
+            if (scheme === "file") {
+                this.settings.targets.local = {
+                    type: "local",
+                    path: path,
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                    type: "local",
+                    path: path,
+                };
+            } else if (scheme === "nfs") {
+                this.settings.targets.nfs = {
+                    type: scheme,
+                    // on read full path is used as export
+                    export: path,
+                    server: server,
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                    export: path,
+                    server: server,
+                };
+            } else {
+                this.settings.targets[scheme] = {
+                    type: scheme,
+                    path: path,
+                    server: server,
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                    path: path,
+                    server: server,
+                };
+                // sshkey is used by ssh and sftp/scp
+                if ("KDUMP_SSH_IDENTITY" in this.settings._internal) {
+                    this.settings.targets[scheme].sshkey =
+                        this.settings._internal.KDUMP_SSH_IDENTITY.value;
+                }
+            }
```
These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+            }
+        }
+
+        // default to local if no target configured
+        if (Object.keys(this.settings.targets).length === 0)
+            this.settings.targets.local = { type: "local" };
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+        if (Object.keys(this.settings.targets).length === 0)
+            this.settings.targets.local = { type: "local" };
+
+        this.settings.compression.enabled = (
+            !("KDUMP_DUMPFORMAT" in this.settings._internal) ||
+            // TODO: what about other compression formats (lzo, snappy)?
+            this.settings._internal.KDUMP_DUMPFORMAT.value === "compressed"
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+     * make sure setting exists if value is not empty
+     * don't delete existing settings
+     */
+    _updateSetting(settings, key, value) {
+        if (key in settings._internal) {
+            settings._internal[key].value = value;
+        } else {
+            if (value)
+                settings._internal[key] = { value: value };
+        }
+    }
```
These 8 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+     * this.settings = current state from file
+     * settings = in-memory state from UI
+     */
+    _persistSettings(settings) {
+        // target
+        if (Object.keys(settings.targets).length > 0) {
+            const target = Object.values(settings.targets)[0];
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+        if (Object.keys(settings.targets).length > 0) {
+            const target = Object.values(settings.targets)[0];
+
+            if ("sshkey" in target)
+                this._updateSetting(settings, "KDUMP_SSH_IDENTITY", target.sshkey);
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+            if ("sshkey" in target)
+                this._updateSetting(settings, "KDUMP_SSH_IDENTITY", target.sshkey);
+
+            let savedir;
+            // default for empty path (except nfs, see below)
+            let path = target.path || "/var/crash";
+            if (path && !path.startsWith("/"))
+                path = "/" + path;
+            if (target.type === "local") {
+                savedir = "file://" + path;
+            } else if (target.type === "nfs") {
+                // override empty path default as nfs path is merged into export on read
+                if (!target.path)
+                    path = "";
+                let exprt = target.export;
+                if (!exprt.startsWith("/"))
+                    exprt = "/" + exprt;
+                savedir = "nfs://" + target.server + exprt + path;
+            } else {
+                savedir = target.type + "://" + target.server + path;
+            }
+            this._updateSetting(settings, "KDUMP_SAVEDIR", savedir);
+        }
+        // compression
+        if (this.settings.compression.enabled != settings.compression.enabled) {
+            if (settings.compression.enabled) {
+                this._updateSetting(settings, "KDUMP_DUMPFORMAT", "compressed");
+            } else {
+                this._updateSetting(settings, "KDUMP_DUMPFORMAT", "ELF");
+            }
+        }
+        return settings;
+    }
```
These 30 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+
+    /* generate the config file from raw text and settings
+     */
+    _generateConfig(settings) {
+        settings = this._persistSettings(settings);
```
These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+    _generateConfig(settings) {
+        settings = this._persistSettings(settings);
+
+        const lines = this._lines.slice(0);
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+
+        const lines = this._lines.slice(0);
+
+        // we take the lines from our last read operation and modify them with the new settings
+        Object.keys(settings._internal).forEach((key) => {
+            const entry = settings._internal[key];
```
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+        Object.keys(settings._internal).forEach((key) => {
+            const entry = settings._internal[key];
+
+            let value = entry.value !== undefined ? entry.value : "";
+            // quote what was quoted before + empty values + multi-word values
+            if (entry.quoted || value === "" || value.includes(" "))
+                value = '"' + value + '"';
+            let line = key + "=" + value;
+            if (entry.suffix)
+                line = line + " " + entry.suffix;
+            // this might be a new entry
+            if (!(key in this._originalSettings)) {
+                lines.push(line);
+                return;
+            }
+            // otherwise edit the old line
+            const origEntry = this._originalSettings[key];
+            lines[origEntry.index] = line;
+        });
```
These 16 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+            lines[origEntry.index] = line;
+        });
+
+        // make sure file ends with a newline
+        if (lines[lines.length - 1] !== "")
+            lines.push("");
+        return lines.join("\n");
+    }
```
These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-094857-1fd368e9-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt commented:
Note that your first two commits will also be squashed (test and code ought to go together); so as long as the end result is only one commit, you can push fixup commits as you like. Most of us do force-pushes (github generates interdiffs just fine), but that's a matter of preference.

I am not sure why that test fails, probably the console API is asynchronous? I suggest converting it to a `wait()` call. Check `git grep wait.*lambda test/verify/` for plenty of existing examples.

skazi0 commented:
I converted the JS log tests as suggested.

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+     * don't delete existing settings
+     */
+    _updateSetting(settings, key, value) {
+        if (key in settings._internal) {
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+    _updateSetting(settings, key, value) {
+        if (key in settings._internal) {
+            settings._internal[key].value = value;
+        } else {
+            if (value)
+                settings._internal[key] = { value: value };
+        }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+                    path = "";
+                let exprt = target.export;
+                if (!exprt.startsWith("/"))
+                    exprt = "/" + exprt;
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+        Object.keys(settings._internal).forEach((key) => {
+            const entry = settings._internal[key];
+
+            let value = entry.value !== undefined ? entry.value : "";
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+            if (entry.suffix)
+                line = line + " " + entry.suffix;
+            // this might be a new entry
+            if (!(key in this._originalSettings)) {
+                lines.push(line);
+                return;
+            }
```
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on pkg/kdump/config-client-suse.js:
```diff
@@ -0,0 +1,265 @@
+
+        // make sure file ends with a newline
+        if (lines[lines.length - 1] !== "")
+            lines.push("");
```
This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-17634-20220902-133021-0d93bc50-fedora-36-devel//Coverage/lcov/github-pr.diff.gcov.html)

martinpitt reviewed (approved):
Thanks! This looks fine to me now. I'll squash everything together on landing.
